File: 3201-distribute-candies-among-children-ii/distribute-candies-among-children-ii.java
class Solution {
    public long distributeCandies(int n, int limit) {
        // Sử dụng nguyên lý bao gồm - loại trừ (Inclusion-Exclusion)
        // Bước 1: Tổng số bộ ba không âm (a + b + c = n)
        // Bước 2: Trừ đi các trường hợp có ít nhất 1 biến vượt quá limit
        // Bước 3: Cộng lại các trường hợp có 2 biến vượt limit (bị trừ hai lần ở bước 2)
        // Bước 4: Trừ lại trường hợp cả 3 biến đều vượt limit (bị cộng dư)

        return combCount(n) // Tất cả các bộ (a + b + c = n) không giới hạn
                - 3 * combCount(n - (limit + 1)) // Trừ các trường hợp có 1 biến > limit
                + 3 * combCount(n - 2 * (limit + 1)) // Cộng lại các TH có 2 biến > limit (bị trừ 2 lần)
                - combCount(n - 3 * (limit + 1)); // Trừ lại TH cả 3 biến đều > limit (bị cộng 3 lần)
    }

    // Hàm tính số cách chọn bộ ba số không âm (a + b + c = sum)
    // Dựa vào tổ hợp C(sum + 2, 2) = (sum + 2)(sum + 1) / 2
    private long combCount(long sum) {
        if (sum < 0)
            return 0; // Nếu sum âm thì không có cách nào hợp lệ
        return (sum + 2) * (sum + 1) / 2;
    }
}

File: 3201-distribute-candies-among-children-ii/distribute-candies-among-children-ii.go
func distributeCandies(n int, m int) int64 {
	// Bước 1: Tính tổng số bộ (a, b, c) không âm thỏa a + b + c = n
	// Công thức tổ hợp: C(n + 2, 2) = (n + 2)(n + 1)/2
	res := int64(n+2) * int64(n+1) / 2;

	// Bước 2: Áp dụng nguyên lý bao gồm - loại trừ để loại bỏ trường hợp có biến > m
	for i := 1; i <= 3; i++ {
		// Tổng còn lại sau khi trừ i biến vượt quá m
		rem := int64(n) - int64(i)*(int64(m)+1);
		if rem < 0 {
			break;
		}

		// Tính số bộ không âm có tổng = rem
		val := (rem + 2) * (rem + 1) / 2;

		// Số cách chọn i biến vượt limit: C(3, i) = 3 (i=1,2) hoặc 1 (i=3)
		c := int64(1);
		if i < 3 {
			c = 3;
		}

		// Bao gồm - loại trừ
		if i%2 != 0 {
			res -= c * val;
		} else {
			res += c * val;
		}
	}

	return res;
}